google/cloud/internal/parse_rfc3339.cc

```diff
@@ -14,218 +14,25 @@
-    auto constexpr kExpectedOffsetFields = 2;
-    if (count != kExpectedOffsetFields || pos != kExpectedOffsetWidth) {
-      ReportError(timestamp, "Invalid timezone offset, expected [+-]HH:MM.");
-    }
-    if (hours < 0 || hours >= kHoursInDay) {
-      ReportError(timestamp, "Out of range offset hour.");
-    }
-    if (minutes < 0 || minutes >= kMinutesInHour) {
-      ReportError(timestamp, "Out of range offset minute.");
-    }
-    buffer += pos;
-    using std::chrono::duration_cast;
-    if (positive) {
-      return duration_cast<std::chrono::seconds>(std::chrono::hours(hours) +
-                                                 std::chrono::minutes(minutes));
-    }
-    return duration_cast<std::chrono::seconds>(-std::chrono::hours(hours) -
-                                               std::chrono::minutes(minutes));
-  }
-  if (buffer[0] != 'Z' && buffer[0] != 'z') {
-    ReportError(timestamp, "Invalid timezone offset, expected 'Z' or 'z'.");
-  }
-  ++buffer;
-  return std::chrono::seconds(0);
-}
-
-#include "google/cloud/internal/diagnostics_pop.inc"
-}  // anonymous namespace
+#include "absl/strings/ascii.h"
+#include "absl/time/time.h"
 
 namespace google {
 namespace cloud {
 inline namespace GOOGLE_CLOUD_CPP_NS {
 namespace internal {
 std::chrono::system_clock::time_point ParseRfc3339(
     std::string const& timestamp) {
-  // TODO(#530) - dynamically change the timezone offset.
-  // Because this computation is a bit expensive, assume the timezone offset
-  // does not change during the lifetime of the program.  This function takes
-  // the current time, converts to UTC and then convert back to a time_t.  The
-  // difference is the offset.
-  static std::chrono::seconds const kLocalTimeOffset = []() {
-    auto now = std::time(nullptr);
-    std::tm lcl;
-    // The standard C++ function to convert time_t to a struct tm is not thread
-    // safe (it holds global storage), use some OS specific stuff here:
-#if _WIN32
-    gmtime_s(&lcl, &now);
-#else
-    gmtime_r(&now, &lcl);
-#endif  // _WIN32
-    lcl.tm_isdst = 0;  // Force use of the "standard" UTC offset.
-    return std::chrono::seconds(mktime(&lcl) - now);
-  }();
-
-  char const* buffer = timestamp.c_str();
-  auto time_point = ParseDateTime(buffer, timestamp);
-  auto fractional_seconds = ParseFractionalSeconds(buffer, timestamp);
-  std::chrono::seconds offset = ParseOffset(buffer, timestamp);
-
-  if (buffer[0] != '\0') {
-    ReportError(timestamp, "Additional text after RFC 3339 date.");
-  }
-
-  time_point += fractional_seconds;
-  time_point -= offset;
-  time_point -= kLocalTimeOffset;
-  return time_point;
+  std::string err;
+  absl::Time t;
+  // TODO(#5593): Once we take a dep on Abseil >= Sept 2020 LTS, we can remove
+  // this call to `absl::AsciiStrToUpper()`, because then `absl::ParseTime()`
+  // will handle `T` and `Z` in upper or lower case.
+  auto const upper = absl::AsciiStrToUpper(timestamp);
```

Comment:
> if the only reason we were waiting on the new Abseil release was because of the case sensitivity issues, I think I'd rather take this code change now, even at the cost of the use of to-upper. Thoughts?

Eliminating the need for the "to upper" by updating the Abseil dep was all my version was waiting for.  It always seemed to be "any day now", but if you think that day is still far off, so be it.

Reply:
I do think it's worth taking this change now so we get the majority of the refactoring improvement, and the only price is a single extra "to upper". I'd rather wait on eliminating that to upper than wait on the rest of the refactoring.